cilium-operator crashes when PodIPPool is shrunk

### Is there an existing issue for this?

- [x] I have searched the existing issues

### Version

equal or higher than v1.18.0 and lower than v1.19.0

### What happened?


`updateCIDRSets` iterates over the existing slice of CIDRAllocator objects while deleting elements from the same slice:
```
for i, oldCIDR := range cidrSets {
        exists := slices.ContainsFunc(newCIDRs, oldCIDR.IsClusterCIDR)
        if !exists {
                cidrSets[i] = nil                                          // poison the slot
                cidrSets = slices.Delete(cidrSets, i, i+1) // shrink slice
        }
}
```
On the next iteration the loop fetches the nil item into `oldCIDR` and calls `oldCIDR.IsClusterCIDR`, resulting in the panic observed.

### How can we reproduce the issue?

Easy reproduction:
```
kubectl apply -f <(echo 'apiVersion: cilium.io/v2alpha1
kind: CiliumPodIPPool
metadata: { name: demo-pool }
spec:
  ipv4:
    cidrs: ["10.0.0.0/24", "10.0.1.0/24"]
    maskSize: 32
') 
kubectl patch cpip demo-pool --type merge -p '{"spec":{"ipv4":{"cidrs":["10.0.1.0/24"]}}}'
```

It only works when removing the penultimate or earlier item in the cidrs list.

### Cilium Version

1.18.0

### Kernel Version

Darwin Kernel Version 24.5.0

### Kubernetes Version

1.33

### Regression

_No response_

### Sysdump

_No response_

### Relevant log output

```shell
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x231799c]

goroutine 387 [running]:
github.com/cilium/cilium/pkg/ipam/allocator/multipool.(*PoolAllocator).updateCIDRSets(0x4000d96bc0, 0x0, {0x4000a5fba0, 0x2, 0x2}, {0x40000cbae0, 0x1, 0x1d0f758?}, 0x20)
	/go/src/github.com/cilium/cilium/pkg/ipam/allocator/multipool/pool_allocator.go:159 +0x26c
github.com/cilium/cilium/pkg/ipam/allocator/multipool.(*PoolAllocator).UpsertPool(0x4000d96bc0, {0x4000eb9b20, 0x9}, {0x4001114e90, 0x1, 0x1}, 0x20, {0x0, 0x0, 0x0}, ...)
	/go/src/github.com/cilium/cilium/pkg/ipam/allocator/multipool/pool_allocator.go:234 +0x23c
github.com/cilium/cilium/pkg/ipam/allocator/multipool.(*Allocator).UpsertPool(0x4000bb7f60, {0x0?, 0x4000635590?}, 0x400078eea0)
	/go/src/github.com/cilium/cilium/pkg/ipam/allocator/multipool/allocator.go:63 +0x2c8
github.com/cilium/cilium/operator/watchers.StartIPPoolAllocator.func1()
	/go/src/github.com/cilium/cilium/operator/watchers/cilium_podippool.go:163 +0x198
created by github.com/cilium/cilium/operator/watchers.StartIPPoolAllocator in goroutine 249
	/go/src/github.com/cilium/cilium/operator/watchers/cilium_podippool.go:154 +0x144
```

### Anything else?

_No response_

### Cilium Users Document

- [ ] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [x] I agree to follow this project's Code of Conduct

--- a/pkg/ipam/allocator/multipool/pool_allocator.go
+++ b/pkg/ipam/allocator/multipool/pool_allocator.go
@@ -156,6 +156,9 @@ func (p *PoolAllocator) updateCIDRSets(isV6 bool, cidrSets []cidralloc.CIDRAlloc
 
 	// delete CIDR set for CIDRs not present in the new CIDRs
 	for i, oldCIDR := range cidrSets {
+		if oldCIDR == nil {
+			continue
+		}
 		exists := slices.ContainsFunc(newCIDRs, oldCIDR.IsClusterCIDR)
 		if !exists {
 			cidrSets[i] = nil

--- a/pkg/ipam/allocator/multipool/pool_allocator_test.go
+++ b/pkg/ipam/allocator/multipool/pool_allocator_test.go
@@ -500,3 +500,28 @@ func Test_addrsInPrefix(t *testing.T) {
 		})
 	}
 }
+
+// TestUpdateCIDRSets_ShrinkPool ensures that shrinking a pool does not
+// trigger a nil dereference in updateCIDRSets.
+func TestUpdateCIDRSets_ShrinkPool(t *testing.T) {
+	p := NewPoolAllocator(hivetest.Logger(t))
+
+	// Initial pool with two IPv4 CIDRs
+	err := p.UpsertPool("shrink-test",
+		[]string{"10.0.0.0/16", "10.1.0.0/16"}, 24,
+		nil, 0,
+	)
+	assert.NoError(t, err)
+
+	pool := p.pools["shrink-test"]
+
+	// Shrink pool to a single CIDR
+	newCIDRs := []netip.Prefix{netip.MustParsePrefix("10.1.0.0/16")}
+
+	assert.NotPanics(t, func() {
+		updated, err := p.updateCIDRSets(false, pool.v4, newCIDRs, 24)
+		assert.NoError(t, err)
+		assert.Len(t, updated, 1)
+		assert.True(t, updated[0].IsClusterCIDR(newCIDRs[0]))
+	})
+}

